File: 199-binary-tree-right-side-view/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

    

class Solution {

    void preOrder(TreeNode root , int lavel , List<Integer> result) {
        if(root==null) {
            return ;
        }

        if(result.size() < lavel) {
            result.add(root.val);
        }

        preOrder(root.right , lavel+1 , result);
        //backtrack
        preOrder(root.left , lavel+1 , result);
    }

    public List<Integer> rightSideView(TreeNode root) {
        // Using DFS

        List<Integer> result = new ArrayList<>();

        preOrder(root , 1 , result);

        return result;
    }
}

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[int]
        """
        result = []
        if not root:
            return result

        queue = deque()
        queue.append(root)

        while queue:
            level_size = len(queue)
            for i in range(level_size):
                node = queue.popleft()
                
                # if it's the last node in this level, add to result
                if i == level_size - 1:
                    result.append(node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        
        return result
